### PR TITLE
JW-270 Added support for registering physical interfaces, vhosts, gat…

### PR DESCRIPTION
**Pull request by Attil**
…eways

### CODE REVIEW

File: windows/vr_host.c [hunk 1]
```diff
@@ -7,6 +7,11 @@
 #include "vr_windows.h"
 #include "vrouter.h"
 
+#define WIN_VHOST_PORTID 1
+#define WIN_VHOST_NICID 0
+#define WIN_PHYSICAL_PORTID 2
+#define WIN_PHYSICAL_NICID 0
```
Comment (michal-clapinski) [inline windows/vr_host.c]: Are you sure about those values? Can't they be dynamically assigned?

File: windows/vr_host.c [hunk 2]
```diff
@@ -1033,26 +1038,58 @@ win_soft_reset(struct vrouter *router)
...
+        if (assoc->port_id != 0 || assoc->nic_index != 0) { // There was already an oid request so you can get port_id, nic_index in the assoc field, so both name_map and ids_map should be updated
+            vif->vif_port = assoc->port_id;
+            vif->vif_nic = assoc->nic_index;
 
-    if (assoc->port_id != 0 || assoc->nic_index != 0) { // There was already an oid request so you can get port_id, nic_index in the assoc field, so both name_map and ids_map should be updated
-        vif->vif_port = assoc->port_id;
-        vif->vif_nic = assoc->nic_index;
+            assoc = vr_get_assoc_ids(assoc->port_id, assoc->nic_index);
+            assoc->interface = vif;
+        }
```
Comment (michal-clapinski) [inline windows/vr_host.c]: Maybe assert instead of `if`?

File: windows/vrouter_mod.c [hunk 3]
```diff
@@ -133,6 +133,17 @@ debug_print_net_buffer(PNET_BUFFER nb, const char *prefix)
 static NDIS_STATUS
 UpdateNics(PNDIS_SWITCH_NIC_PARAMETERS Nic, BOOLEAN connect)
 {
+    /* The structure is as follows:
+     * PortID 1: Nic 0: vhost
+     * PortID 2: Nic 0: Default external interface
+     * PortID 2: Nic 1+: Specific external interfaces
+     * PortID 3+: Containers/VMs
+     */
+    if (Nic->PortId <= 2)
```
Comment (michal-clapinski) [inline windows/vrouter_mod.c]: You should use a define instead

File: windows/vrouter_mod.c [hunk 4]
```diff
@@ -913,8 +933,16 @@ SxExtStartNetBufferListsIngress(
         NDIS_SWITCH_PORT_ID source_port = fwd_detail->SourcePortId;
         NDIS_SWITCH_NIC_INDEX source_nic = fwd_detail->SourceNicIndex;
         windows_host.hos_printf("%s: port %d and interface id %d\n", __func__, source_port, source_nic);
-        struct vr_assoc *assoc_entry = vr_get_assoc_ids(source_port, source_nic);
-        struct vr_interface *vif = (assoc_entry ? assoc_entry->interface : NULL);
+
+        struct vr_interface *vif;
+        if (source_port == 1)
+            vif = router->vr_host_if;
```
Comment (michal-clapinski) [inline windows/vrouter_mod.c]: Wouldn't it just be simpler to add those to assoc map?

File: windows/vrouter_mod.c [hunk 5]
```diff
@@ -387,6 +399,14 @@ SxExtRestartSwitch(
 
     struct vr_switch_context *ctx = (struct vr_switch_context *)ExtensionContext;
 
+    PNDIS_SWITCH_NIC_ARRAY array;
+    SxLibGetNicArrayUnsafe(Switch, &array);
+
+    for (unsigned i = 0; i < array->NumElements; i++){
+        PNDIS_SWITCH_NIC_PARAMETERS element = NDIS_SWITCH_NIC_AT_ARRAY_INDEX(array, i);
+        UpdateNics(element, TRUE);
```
Comment (michal-clapinski) [inline windows/vrouter_mod.c]: Why did you add these lines? `UpdateNics` doesn't do anything for gateway and physical.

File: windows/vr_host.c [hunk 6]
```diff
@@ -1033,26 +1038,58 @@ win_soft_reset(struct vrouter *router)
...
+        vif->vif_nic = WIN_VHOST_NICID;
+        break;
+    }
 
-        vif_attach(vif);
+    default:
+    {
+        ASSERTMSG("Non-supported VIF type", FALSE);
     }
+    }
```
Comment (michal-clapinski) [inline windows/vr_host.c]: You've got a weird thing going with those brackets and indentations. Also you don't need brackets for cases in which you don't create variables.

File: windows/vr_host.c (hunk 1 above)
Comment (Attil) [inline windows/vr_host.c]: There are a few ways to go around this:
1. Like it is implemented right now.
1. One can determine physical interface by type (This doesn't work for vHost, as both vHost and any virtual interface have the same types)
1. By Friendly Name, this is an option I dislike, mainly because friendly names shouldn't be really used like this.
1. Any other?

File: windows/vr_host.c (hunk 2 above)
Comment (Attil) [inline windows/vr_host.c]: This is not a sanity check, we should support both cases, ie. OID request comes first and Sandesh message comes first.

Comment (Attil) [inline windows/vr_host.c]: This is not a sanity check, we should support both cases, ie. OID request comes first and Sandesh message comes first.

File: windows/vr_host.c (hunk 6 above)
Comment (Attil) [inline windows/vr_host.c]: Yeah, I thought about it. I need brackets when declaring a new variable to get a new scope, but the code looks strange when one `case` has brackets and the other doesn't. What do you suggest?

File: windows/vrouter_mod.c (hunk 5 above)
Comment (Attil) [inline windows/vrouter_mod.c]: Because in the future it can something, the fact it doesn't is implementation detail from API point of view.
